### PR TITLE
Fix stats and add latency avg max min as well latest rates

### PR DESCRIPTION
**Pull request by speeddragon**
This PR fixes the stats information and also adds the following metrics:
* Latency (`avg`, `max`, `min`)
* Latest rate request since last update.

### CODE REVIEW

File: lib/load/ws_handler.ex [hunk 1]
```diff
@@ -1,21 +1,30 @@
...
     {:cowboy_websocket, req, state}
   end
 
+  @impl true
+  def websocket_init(state) do 
+    :pg.join(WS, self())
+
+    {:ok, state}
+  end
+
```
Comment (domhaobaobao) [inline lib/load/ws_handler.ex:22]: good catch :+1:

Comment (domhaobaobao) [inline lib/load/ws_handler.ex:22]: good catch :+1:

File: lib/load.ex [hunk 2]
```diff
@@ -25,6 +29,13 @@ defmodule Load do
     end)
   end
 
+  def stop() do
+    DynamicSupervisor.which_children(Load.Worker.Supervisor)
+    |> Enum.each(fn {_id, pid, :worker, _} ->
+      send(pid, :exit)
+    end)
+  end
```
Comment (domhaobaobao) [inline lib/load.ex:37]: I think this only stops local children, I would rather go through scale down to implement stop as this has an effect only on the local node.

File: lib/load/stats.ex [hunk 3]
```diff
@@ -8,14 +8,16 @@ defmodule Stats do
     last_ms: 0, # last time stats were collected
     requests: 0,
     succeeded: 0,
-    failed: 0
+    failed: 0,
+    latencies: []
```
Comment (domhaobaobao) [inline lib/load/stats.ex:12]: latencies were intentionally left out as the only latency that makes sense in a p2p traversal is the one emerging from the collector

Comment (domhaobaobao) [reply lib/load/stats.ex:12]: that said it helps to have them, for other use cases (i.e. test a web server)

File: lib/load/worker.ex [hunk 4]
```diff
@@ -43,16 +43,29 @@ defmodule Load.Worker do
...
+
+    Process.send_after(self(), :run, (interval_ms - latency_ms))
+    
     {:noreply, state}
   end
 
-  def hit(target, headers, payload, state) do
+  def handle_info(:exit, state) do
+    {:stop, :normal, state}
+  end
```
Comment (domhaobaobao) [inline lib/load/worker.ex:66]: like above, this is not the right approach

File: lib/load/ws_handler.ex [hunk 5]
```diff
@@ -55,15 +69,18 @@ defmodule Load.WSHandler do
     {:reply, {:text, Jason.encode!(%{stats: stats})}, state}
   end
 
-  @impl true
+  def websocket_info(:ping, state) do 
+    {:ok, state}
+  end
```
Comment (domhaobaobao) [inline lib/load/ws_handler.ex:74]: should be the server sending pings and receiving back pongs, should we remove this?

**Review by domhaobaobao — CHANGES_REQUESTED** [review]

File: lib/load.ex (hunk 2 above)
Comment (domhaobaobao) [reply lib/load.ex:37]: please implement via scale(0)

File: lib/load/worker.ex (hunk 4 above)
Comment (domhaobaobao) [reply lib/load/worker.ex:66]: let us use scale(0)

File: lib/load.ex (hunk 2 above)
Comment (domhaobaobao) [reply lib/load.ex:37]: def stop, do: scale(0)

File: lib/load/worker.ex [hunk 6]
```diff
@@ -43,16 +43,29 @@ defmodule Load.Worker do
...
+
+    latency_ms = (System.monotonic_time - start_time) / 1000000 |> trunc()
+
+    Logger.debug("Request latency: #{latency_ms} ms")
+
+    state = %{state | latencies: [latency_ms] ++ latencies}
     |> Stats.maybe_update()
-    Process.send_after(self(), :run, interval_ms)
+
+    Process.send_after(self(), :run, (interval_ms - latency_ms))
```
Comment (domhaobaobao) [inline lib/load/worker.ex:59]: I would restore the original here, this assumes latency is always smaller than interval. This is dangerous, and a bit unnecessary

File: lib/load.ex (hunk 2 above)
Comment (hauleth) [reply lib/load.ex:37]: Why not just stop supervisor. This should stop all children as well.

**Pull request closed without merging**